enzyme/Enzyme/AdjointGenerator.h

```diff
@@ -9785,15 +9872,24 @@ class AdjointGenerator
           IRBuilder<> Builder2(&call);
           getForwardBuilder(Builder2);
           auto origfree = orig->getArgOperand(0);
+          auto newfree = gutils->getNewFromOriginal(orig->getArgOperand(0));
           auto tofree = gutils->invertPointerM(origfree, Builder2);
-          if (tofree != origfree) {
-            SmallVector<Value *, 2> args = {tofree};
-            CallInst *CI = Builder2.CreateCall(orig->getFunctionType(),
-                                               orig->getCalledFunction(), args);
-            CI->setAttributes(orig->getAttributes());
-          }
+
+          Function *free = getOrInsertCheckedFree(
+              *orig->getModule(), orig->getFunctionType(),
+              orig->getCalledFunction(), newfree->getType(),
+              orig->getAttributes(), gutils->getWidth());
+
+          SmallVector<Value *, 3> args;
+          args.push_back(newfree);
+
+          auto rule = [&args](Value *tofree) { args.push_back(tofree); };
+          applyChainRule(Builder2, rule, tofree);
+
+          Builder2.CreateCall(free->getFunctionType(), free, args);
```

Comment:
Can you add the debug info of the original function, and also add the calling convention?

Reply:
done